File: app/controllers/user_controller.rb
class UserController < ApplicationController
  def sign_in
  end

  def sign_up
  end

  def create_user
    permitted_params = params.permit(:firstName, :lastName, :email, :password, :confirmPassword)

    user_data = {
      first_name: permitted_params[:firstName],
      last_name: permitted_params[:lastName],
      email: permitted_params[:email],
      password: permitted_params[:password],
      password_confirmation: permitted_params[:confirmPassword]
    }

    user = User.new(user_data)
    result = {
      success: user.save,
      errors: user.errors.full_messages
    }

    render json: { result: result }
  end
end

File: app/javascript/controllers/user_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["firstName", "lastName", "email", "password", "confirmPassword"]
  
  connect() {
    console.log("USER CONTROLLER STIMULUS CONNECTED")
  }

  async signUp(event) {
    // Prevents redirecting on form submission
    event.preventDefault();

    // Update button state
    const button = event.currentTarget
    button.classList.add("disabled-button")
    button.innerText = "Submitting..."

    // Sends post request for user creation
    const response = await this.#createUser()
    const result = await response.json()
    debugger
  }

  #createUser() {
    const url = "/user/create_user"
    return fetch(url, {
      method: "POST",
      headers: { 
        "Content-Type": "application/json"
      },
      body: JSON.stringify(this.#buildUserData())
    })
  }

  #buildUserData() {
    return {
      firstName: this.firstNameTarget.value,
      lastName: this.lastNameTarget.value,
      email: this.emailTarget.value,
      password: this.passwordTarget.value,
      confirmPassword: this.confirmPasswordTarget.value
    }
  }
}
